packages/flutter_tools/lib/src/isolated/devtools_launcher.dart

```diff
@@ -15,47 +19,94 @@ import '../resident_runner.dart';
 /// a devtools dep in google3.
 class DevtoolsServerLauncher extends DevtoolsLauncher {
   DevtoolsServerLauncher({
+    @required ProcessManager processManager,
+    @required String pubExecutable,
     @required Logger logger,
-  }) : _logger = logger;
+  })  : _processManager = processManager,
+        _pubExecutable = pubExecutable,
+        _logger = logger;
 
+  final ProcessManager _processManager;
+  final String _pubExecutable;
   final Logger _logger;
 
-  io.HttpServer _devtoolsServer;
+  io.Process _devToolsProcess;
+  Uri _devToolsUri;
+
+  static final RegExp _serveDevToolsPattern =
+      RegExp(r'Serving DevTools at ((http|//)[a-zA-Z0-9:/=_\-\.\[\]]+)');
 
   @override
-  Future<void> launch(Uri observatoryAddress) async {
+  Future<void> launch(Uri vmServiceUri, {bool openInBrowser = false}) async {
+    if (_devToolsProcess != null && _devToolsUri != null) {
+      // DevTools is already running.
+      if (openInBrowser) {
+        await Chrome.start(<String>[_devToolsUri.toString()]);
+      }
+      return;
+    }
+
     try {
-      await serve();
-      await devtools_server.launchDevTools(
-        <String, dynamic>{
-          'reuseWindows': true,
-        },
-        observatoryAddress,
-        'http://${_devtoolsServer.address.host}:${_devtoolsServer.port}',
-        false,  // headless mode,
-        false,  // machine mode
-      );
+      final io.ProcessResult _devToolsActivateProcess = await _processManager.run(<String>[
```

Comment:
Is there any way to check whether we have to run this every time unconditionally? What sort of delay does it introduce?

Reply:
To clarify, is that 4.5 seconds on every run, or only if the package has not already been activated/needs an update?

Reply:
4.5 seconds on each run. If this is too severe, we could instead fail and print some help text "package:devtools not activated. Please run `flutter pub global activate devtools`, and try again". However, if we can make the change to launch the the devtools server at app run, this should only be run once. Then all further calls to launch should return the already-running server, not spin up a new one.

Reply:
Risking slowing down app launch is a bit worse. Let's wrap this in a `Logger.startProgress`: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/logger.dart#L126. There are examples of its use in several places in the code to crib from.

Additionally, is there an issue we can note here in a comment that tracks the slow run time for pub when being asked to do a no-op `global activate`?

Reply:
added `logger.startProgress` and linked to https://github.com/dart-lang/pub/issues/2791